pyteal/ast/scratch.py

```diff
@@ -59,41 +59,92 @@ def load(self, type: TealType = TealType.anytype) -> "ScratchLoad":
         """
         return ScratchLoad(self, type)
 
+    def index(self) -> "ScratchIndex":
+        return ScratchIndex(self)
+
     def __repr__(self):
         return "ScratchSlot({})".format(self.id)
 
     def __str__(self):
         return "slot#{}".format(self.id)
 
+
+ScratchSlot.__module__ = "pyteal"
+
+
+class ScratchIndex(Expr):
+    def __init__(self, slot: ScratchSlot):
+        super().__init__()
+        self.slot = slot
+
+    def __str__(self):
+        return "(ScratchIndex {})".format(self.slot)
+
     def __hash__(self):
-        return hash(self.id)
+        return hash(self.slot.id)
 
+    def type_of(self):
+        return TealType.uint64
 
-ScratchSlot.__module__ = "pyteal"
+    def has_return(self):
```

Comment:
@tzaffi Clarification question:  How have folks been treating membership for `LeafExpr`?

* I see `LeafExpr` exists to delineate `Expr`s that do _not_ always return from the current subroutine or program.
* Based on `ScratchIndex.has_return` implementation, it seems `ScratchIndex` is eligible.
  * Aside:  The same observation holds for other classes in `scratch.py` (e.g. `ScratchLoad`).
* I don't see references to `LeafExpr` for controlling program behavior.  
  * I'm a little surprised because I assumed the type marker can be used to control behavior.
  * Nevertheless, I'm _imagining_ `LeafExpr` as a member of an algebraic data type (ADT).  And in my experience, ADTs help maintainers gauge the shape of supported behaviors.
* In summary - _If_ `ScratchIndex` is eligible to be a `LeafExpr`, I don't feel strongly today about making a change.  I imagine having a stronger opinion when I develop deeper PyTeal intuitions.

Reply:
Worthy of discussion

Reply:
I don't think we directly discussed the topic verbally.

* Although I'm interested in a discussion eventually, I'm not looking to hold the PR.  So, I'll resolve the thread.
* The verbal discussion highlighted a bias towards `Expr` usage rather than specific `Expr` subtypes.  Indirectly, I imagine that bias may explain why `LeafExpr` is not used more pervasively.